src/game/counting.rs

```diff
@@ -0,0 +1,276 @@
+use crate::game::cards::Card;
+use crate::game::scoring::{score_run, Combination, CombinationKind, Score};
+use std::error::Error;
+
+/// Calculates the score during the counting phase where played_cards have already been played and card is now played
+///
+///
+pub fn score_counting_cards_played(
+    played_cards: &[Card],
+    card: Card,
+) -> Result<Score, Box<dyn Error>> {
+    let count: i32 = played_cards.iter().map(|c| c.value).sum::<i32>() + card.value;
+    if count > 31 {
+        return Err("invalid card + count > 31".into());
+    }
+
+    let mut score: Score = Score::new();
+    let mut all_cards: Vec<Card> = played_cards.to_vec();
+
+    all_cards.push(card);
+    if count == 15 {
+        score.combinations.push(Combination::new(
+            CombinationKind::Fifteen,
+            all_cards.clone(),
+        ));
+    } else if count == 31 {
+        score.combinations.push(Combination::new(
+            CombinationKind::ThirtyOne,
+            all_cards.clone(),
+        ));
+    }
+
+    //
+    //  check to see if we have matching rank
+    let mut len = played_cards.len();
+    let mut count = 0;
+    if len > 0 {
+        for i in 1..=3 {
+            if i > len {
+                // we need to loop for up to 4 of a kind, but not if we don't have 4 cards...
+                break;
+            }
+            if played_cards[len - i..].iter().all(|c| c.rank == card.rank)
+            // card is the played card passed in
+            {
+                count += 1;
+            } else {
+                break;
+            }
+        }
+    }
+
+    if count > 0 {
+        score.combinations.push(Combination::new(
+            CombinationKind::RankMatch,
+            all_cards.as_slice()[len - count..].to_vec(),
+        ));
+    }
+
+    //
+    //   now look for runs - this is a bit complicated in counting because the run doesn't have to be in order
+    //   e.g. Ace, Five, Four, Two, Three is worth 5 points to whoever played the three.
+    //
+    len = all_cards.len();
+    if len > 2 {
+        for i in 0..all_cards.len() - 1 {
+            // get a slice of the last i cards, sort them, and see if they are a run.
+            // we can't sort the slice because they sort in place and return ()
+            let mut possible_run = all_cards.as_slice()[i..].to_vec();
```

Comment:
This chunk is nice and clear on how it works.

Reply:
thank you! :)